### ISSUE TITLE
Loudspeaker doesn't work with Android 12, Galaxy S21 linphone 4.6 

### ISSUE THREAD

**Issue by mircoiplabs** [BelledonneCommunications/linphone-android#1640]
**Describe the bug**
Loudspeaker doesn't work for some Android devices, like Samsung Galaxy S21, Android 12

**To Reproduce**
1. During a call, click on the speaker button
2. The audio disappear
3. Changing the volume up or down, for this device, restore correctly the audio
4. Some users with different devices hears nothing even when the volume is changed

**Please complete the following information**
 - Device: Samsung Galaxy S21
 - OS: Android 12
 - Version of the SDK 5.1.6

**Comment by Viish** [comment]
Please attach app / SDK logs showing the issue.

**Comment by mircoiplabs** [comment]
We just figured out that the problem may appear if the Multimedia Volume of the phone is zero, in this case the user cannot hear anything with loudspeaker on if he doesn't turns the volume up or down, but if the multimedia volume of the device is more than zero the problem seems to not appear

[log linphone Samsung S21 speaker issue.odt](https://github.com/BelledonneCommunications/linphone-android/files/8396635/log.linphone.Samsung.S21.speaker.issue.odt)

**Comment by Viish** [comment]
Can you attach the logs from the start of the app to the end of the call please?
Just so you know, we are aware of this issue on Samsung devices and we already do a "volume up" followed by a "volume down" to workaround this problem.
Another possible solution is to switch to OpenSLES audio driver (deprecated by Google) instead of AAudio.
And this issue was already opened here: https://github.com/BelledonneCommunications/linphone-android/issues/1531

**Comment by mircoiplabs** [comment]
[log linphone samsung s21 android 12 bug speaker.log](https://github.com/BelledonneCommunications/linphone-android/files/8440555/log.linphone.samsung.s21.android.12.bug.speaker.log)

**Comment by Viish** [comment]
From your logs everything looks fine, workaround is correctly called: `[Audio Manager] Lower & raise audio volume to workaround no sound issue until volume has changed...`
So you can either do the same workaround in your app when you change the output audio device, or switch to OpenSLES.

**Comment by mvdhorst** [comment]
Hi @Viish I've got the same issue. For me the procedure seems to be called, but the volume isn't changed, I also see why:
It uses the volume of the STREAM_MUSIC at the start, not that of the STREAM_VOICE_CALL
`2022-04-26 14:50:14.868 11255-19111/nl.clb.linphone.debug D/AudioStreamInternalPlay_Client: doSetVolume() mStreamVolume * duckAndMuteVolume = 0.000000 * 1.000000 = 0.000000
2022-04-26 14:50:14.868 11255-19111/nl.clb.linphone.debug D/AudioStreamInternal_Client: onEventFromServer - got AAUDIO_SERVICE_EVENT_STARTED
2022-04-26 14:50:14.868 11255-19111/nl.clb.linphone.debug D/AAudioStream: setState(s#40) from 3 to 4
2022-04-26 14:50:14.868 11255-19111/nl.clb.linphone.debug D/AudioStreamInternal_Client: onEventFromServer - AAUDIO_SERVICE_EVENT_VOLUME 0.000000
2022-04-26 14:50:14.868 11255-19111/nl.clb.linphone.debug D/AudioStreamInternalPlay_Client: doSetVolume() mStreamVolume * duckAndMuteVolume = 0.000000 * 1.000000 = 0.000000`

**Comment by Viish** [comment]
@mvdhorst Where do you see that AAudio is using STREAM_MUSIC instead of STREAM_VOICE_CALL?

**Comment by Viish** [comment]
@mvdhorst @mircoiplabs I may have a fix for that anyway (just thought of something).
Can you build a SDK with msaaudio submodule on commit id 36686de4? If not I can send you an APK.

**Comment by mvdhorst** [comment]
I can see that it is using the STREAM_MUSIC, because when I alter the STREAM_MUSIC and then start the call, when I press the loudspeaker, it uses that value.
It's been a long time since i build an SDK, i'm spoiled with your pre compiled SDK. Can you send me an APK?

**Comment by Viish** [comment]
Here you go: https://download.linphone.org/snapshots/android/linphone-android-release-4.7.0-alpha.149+bf6dd38b1.apk

**Comment by mvdhorst** [comment]
@Viish thanks! This solves the problem, audio is now heard, only difference I see now is that you set the volume to maximum directly, when I then change the volume using the buttons, the volume is reset to the wanted voice call volume.
Example: I have my voice call volume set to 1 (low)
I start the speaker mode -> volume is set to max
I press the volume up button, the volume is reset to 2, see small part of logging:
```
2022-04-26 16:28:16.128 20652-20652/? D/AudioStreamInternal_Client: open() original HW burst = 240, minMicros = 2000 => SW burst = 240
2022-04-26 16:28:16.155 20652-15472/? D/AudioStreamInternalPlay_Client: callbackLoop() entering >>>>>>>>>>>>>>>
2022-04-26 16:28:16.155 20652-15472/? D/AudioStreamInternal_Client: onEventFromServer - AAUDIO_SERVICE_EVENT_VOLUME 1.000000
2022-04-26 16:28:16.155 20652-15472/? D/AudioStreamInternalPlay_Client: doSetVolume() mStreamVolume * duckAndMuteVolume = 1.000000 * 1.000000 = 1.000000
2022-04-26 16:28:16.155 20652-15472/? D/AudioStreamInternal_Client: onEventFromServer - got AAUDIO_SERVICE_EVENT_STARTED
2022-04-26 16:28:16.155 20652-15472/? D/AudioStreamInternal_Client: onEventFromServer - AAUDIO_SERVICE_EVENT_VOLUME 1.000000
2022-04-26 16:28:16.155 20652-15472/? D/AudioStreamInternalPlay_Client: doSetVolume() mStreamVolume * duckAndMuteVolume = 1.000000 * 1.000000 = 1.000000
2022-04-26 16:28:29.339 20652-15472/? D/AudioStreamInternal_Client: onEventFromServer - AAUDIO_SERVICE_EVENT_VOLUME 0.125893
2022-04-26 16:28:29.339 20652-15472/? D/AudioStreamInternalPlay_Client: doSetVolume() mStreamVolume * duckAndMuteVolume = 0.125893 * 1.000000 = 0.125893
```

**Comment by Viish** [comment]
Can you attach logs from the app please?

**Comment by Viish** [comment]
Actually it seem there are two volume channels: one for the earpiece and one for the speaker (at least on the Z Fold 2 device I use for testing).

**Comment by Viish** [comment]
Fix has been merged in master & release/5.1

**Comment by mircoiplabs** [comment]
Sorry Viish, it seems the "hackVolume()" method isn't changed in the release/5.1 of the SDK, but only in Master.
It's possible to get this fix in the 5.1 of linphone-sdk?

**Comment by Viish** [comment]
@mircoiplabs the fix is in 5.1 SDK, starting 5.1.28.
